presto-parquet/src/main/java/io/prestosql/parquet/predicate/PredicateUtils.java

```diff
@@ -111,22 +110,26 @@ public static boolean predicateMatches(Predicate parquetPredicate, BlockMetaData
         return statistics.build();
     }
 
-    private static Map<ColumnDescriptor, DictionaryDescriptor> getDictionaries(BlockMetaData blockMetadata, ParquetDataSource dataSource, Map<List<String>, RichColumnDescriptor> descriptorsByPath, TupleDomain<ColumnDescriptor> parquetTupleDomain)
+    private static boolean dictionaryPredicatesMatch(Predicate parquetPredicate, BlockMetaData blockMetadata, ParquetDataSource dataSource, Map<List<String>, RichColumnDescriptor> descriptorsByPath, TupleDomain<ColumnDescriptor> parquetTupleDomain)
     {
         ImmutableMap.Builder<ColumnDescriptor, DictionaryDescriptor> dictionaries = ImmutableMap.builder();
         for (ColumnChunkMetaData columnMetaData : blockMetadata.getColumns()) {
             RichColumnDescriptor descriptor = descriptorsByPath.get(Arrays.asList(columnMetaData.getPath().toArray()));
             if (descriptor != null) {
                 if (isOnlyDictionaryEncodingPages(columnMetaData) && isColumnPredicate(descriptor, parquetTupleDomain)) {
-                    int totalSize = toIntExact(columnMetaData.getTotalSize());
-                    byte[] buffer = new byte[totalSize];
+                    byte[] buffer = new byte[toIntExact(columnMetaData.getTotalSize())];
                     dataSource.readFully(columnMetaData.getStartingPos(), buffer);
-                    Optional<DictionaryPage> dictionaryPage = readDictionaryPage(buffer, columnMetaData.getCodec());
-                    dictionaries.put(descriptor, new DictionaryDescriptor(descriptor, dictionaryPage));
+                    DictionaryDescriptor dictionary = new DictionaryDescriptor(descriptor, readDictionaryPage(buffer, columnMetaData.getCodec()));
+                    //  Early abort, predicate already filters block so no more dictionaries need be read
+                    if (!parquetPredicate.matches(dictionary)) {
+                        return false;
+                    }
+                    dictionaries.put(descriptor, dictionary);
                 }
             }
         }
-        return dictionaries.build();
+        // Attempt full domain check
+        return parquetPredicate.matches(dictionaries.build());
```

Comment:
can this return `true` when each earlier attempt returned `false`?

add explanatory code comment, otherwise it's tempting to replace this line with `return true`

Reply:
Good point, removed the redundant check at the end.